yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+   * to re-occur when resubmitted.
+   *
+   * @param entity entity to post
+   * @return tuple with exactly one of the response or the exception set.
+   */
+  private def postOneEntity(entity: TimelineEntity):
+      (Option[TimelinePutResponse], Option[Exception]) = {
+    domainId.foreach {
+      entity.setDomainId
+    }
+    val entityDescription = describeEntity(entity)
+    logDebug(s"About to POST $entityDescription")
+    _entityPostAttempts.incrementAndGet()
+    try {
+      val response = timelineClient.putEntities(entity)
+      val errors = response.getErrors
+      if (errors.isEmpty) {
+        logDebug(s"entity successfully posted")
+        _entityPostSuccesses.incrementAndGet()
+      } else {
+        // something went wrong.
+        errors.asScala.foreach { err =>
+          _entityPostFailures.incrementAndGet()
+          logError(s"Failed to post $entityDescription\n:${describeError(err)}")
+        }
+      }
+      // whatever the outcome, this request is not re-issued
+      (Some(response), None)
+    } catch {
+      case e: ConnectException =>
+        _entityPostFailures.incrementAndGet
+        logWarning(s"Connection exception submitting $entityDescription", e)
+        (None, Some(e))
+
+      case e: RuntimeException =>
+        // this is probably a retry timeout event; some Hadoop versions don't
+        // rethrow the exception causing the problem, instead raising an RTE
+        _entityPostFailures.incrementAndGet
+        logWarning(s"Runtime exception submitting $entityDescription", e)
+        // same policy as before: retry on these
+        (None, Some(e))
+
+      case e: Exception =>
+        // something else has gone wrong.
+        _entityPostFailures.incrementAndGet
+        logWarning(s"Could not handle history entity: $entityDescription", e)
+        (None, Some(e))
+    }
+  }
+
+  /**
+   * Spin and post until stopped.
+   *
+   * Algorithm.
+   *
+   * 1. The thread waits for events in the _entityQueue until stopped or interrupted
+   * 1. Failures result in the entity being queued for resending, after a delay which grows
+   * linearly on every retry.
+   * 1. Successful posts reset the retry delay.
+   * 1. If the process is interrupted, the loop continues with the `stopFlag` flag being checked.
+   *
+   * To stop this process then, first set the `stopFlag` flag, then interrupt the thread.
+   *
+   * @param stopFlag a flag to set to stop the loop.
+   * @param retryInterval delay in milliseconds for the first retry delay; the delay increases
+   *        by this value on every future failure. If zero, there is no delay, ever.
+   */
+  private def postEntities(stopFlag: AtomicBoolean, retryInterval: Long): Unit = {
+    var lastAttemptFailed = false
+    var currentRetryDelay = retryInterval
+    while (!stopped.get) {
+      try {
+        val head = _entityQueue.take()
+        val (_, ex) = postOneEntity(head)
```

Comment:
The first item in the return value of `postOneEntity` is never used; why return it?